File: solutions/uri/1025/1025.cpp
#include <algorithm>
#include <iostream>
#include <vector>

int bi_search(std::vector< int > v, int n) {
    int begin = 0;
    int end = v.size() - 1;
    int middle = end / 2;

    while (begin < end) {
        if (v.at(middle) < n) {
            begin = middle + 1;
            middle = (end + begin) / 2;
        } else if (v.at(middle) > n || (middle > 0 && v.at(middle - 1) == n)) {
            end = middle - 1;
            middle = (end + begin) / 2;
        } else {
            return middle + 1;
        }
    }

    if (v.at(middle) == n) {
        return middle + 1;
    }
    return -1;
}

int main() {
    int n, q, marble, query, position, c = 0;
    std::vector< int > marbles;

    while (std::cin >> n >> q && n && q) {
        marbles.clear();

        while (n--) {
            std::cin >> marble;
            marbles.push_back(marble);
        }

        sort(marbles.begin(), marbles.end());

        std::cout << "CASE# " << ++c << ":" << std::endl;

        while (q--) {
            std::cin >> query;
            position = bi_search(marbles, query);
            if (position == -1) {
                std::cout << query << " not found" << std::endl;
            } else {
                std::cout << query << " found at " << position << std::endl;
            }
        }
    }
    return 0;
}

File: solutions/uri/1025/1025.clj
(ns main
  (:require
   [clojure.string :as str]))

(defn binary-search [value items]
  (loop [begin 0
         end (-> (count items)
                 (dec))]
    (let [middle (-> (+ begin end)
                     (quot 2))
          items (vec items)]
      (if  (> begin end)
        -1
        (cond
          (< (items middle) value)
          (recur (inc middle)
                 end)

          (or (> (items middle) value)
              (and (> middle 0)
                   (== (items (dec middle)) value)))
          (recur begin
                 (dec middle))

          :else
          (inc middle))))))

(defn read-values [total]
  (loop [reverse-index total
         vals []]
    (if (== reverse-index 0)
      vals
      (recur (dec reverse-index)
             (concat vals [(Integer/parseInt (read-line))])))))


(defn process-print [index marbles queries]
  (printf "CASE# %d:%n" index)
  (loop [queries queries]
    (let [query (first queries)
          position (binary-search query marbles)]
      (if (>= position 0)
        (printf "%d found at %d%n" query position)
        (printf "%d not found%n" query)))
    (when (seq (rest queries))
      (recur (rest queries)))))

(defn main []
  (loop [line (read-line)
         index 1]
    (let [[n q] (->> (str/split line #" ")
                     (map #(Integer/parseInt %)))
          marbles (sort (read-values n))
          queries (read-values q)]
      (when (and (not= n 0) (not= q 0))
        (process-print index marbles queries)
        (recur (read-line) (inc index))))))

(main)
